packages/augur-state/src/db/DB.ts

```diff
@@ -146,76 +150,177 @@ export class DB<TBigNumber> {
+      sequenceIds[dbName] = dbInfo.update_seq.toString();
+    }
+    for (let trackedUser of await this.trackedUsers.getUsers()) {
+      for (let userSpecificEvent of userSpecificEvents) {
+        let dbName = `${this.networkId}-${userSpecificEvent.name}-${trackedUser}`;
+        let dbInfo = await this.userSyncableDatabases[dbName].getInfo();
+        sequenceIds[dbName] = dbInfo.update_seq.toString();
       }
     }
+    return sequenceIds;
+  }
 
-    // Sync user-specific event types
-    // TODO TokensTransferred should comprise all balance changes with additional metadata and with an index on the to party.
-    // Also update topics/indexes for user-specific events once these changes are made to the contracts.
+  /**
+   * Rolls back all blocks from blockNumber onward.
+   * 
+   * TODO Add check to see if any rollbacks failed
+   * 
+   * @param {number} blockNumber Oldest block number to delete
+   */
+  public async rollback(blockNumber: number): Promise<void> {
+    // TODO Fix typedef for previousBlockSequenceIds
+    const previousBlockSequenceIds: any = await this.metaDatabase.find(
+      {
+          selector: { blockNumber: { $gte: blockNumber } },
+          fields: ['_id', 'networkId', 'blockNumber', 'sequenceIds'],
+      }
+    );
+    const parseBlockSequenceIds = JSON.parse(previousBlockSequenceIds.docs[0].sequenceIds);
+
+    let dbRollbackPromises = [];
+    // Perform rollback on SyncableDBs & UserSyncableDBs
+    for (let eventName of genericEventNames) {
+      let dbName = `${this.networkId}-${eventName}`;
+      let previousSequenceId = parseBlockSequenceIds[dbName];
+      let dbInfo = await this.syncableDatabases[dbName].getInfo();
+      if (dbInfo.update_seq > previousSequenceId) {
+        dbRollbackPromises.push(this.syncableDatabases[dbName].rollback(blockNumber, previousSequenceId));
+      }
+    }
     for (let trackedUser of await this.trackedUsers.getUsers()) {
       for (let eventIndex in userSpecificEvents) {
-        this.userSyncableDatabases[`${this.networkId}-${userSpecificEvents[eventIndex].name}-${trackedUser}`].sync(augur, chunkSize, blockstreamDelay, uploadBlockNumber);
-        let updateSeq = await this.userSyncableDatabases[`${this.networkId}-${userSpecificEvents[eventIndex].name}-${trackedUser}`].getUpdateSeq();
-        if (typeof updateSeq !== "undefined") {
-          sequenceIds[`${this.networkId}-${userSpecificEvents[eventIndex].name}-${trackedUser}`] = updateSeq;
+        let dbName = `${this.networkId}-${userSpecificEvents[eventIndex].name}-${trackedUser}`;
+        let previousSequenceId = parseBlockSequenceIds[dbName];
+        let dbInfo = await this.userSyncableDatabases[dbName].getInfo();
+        if (dbInfo.update_seq > previousSequenceId) {
+          dbRollbackPromises.push(this.userSyncableDatabases[dbName].rollback(blockNumber, previousSequenceId));
         }
       }
     }
+    await Promise.all(dbRollbackPromises);
```

Comment:
What happens if any given `dbRollbackPromises` fails? Will it put us in a bad state?

Reply:
That's a good question... we should probably fail and see what failed?

Reply:
Agreed. So if any of these promises gets rejected, we should output the failure and terminate the stop running at that point? Or do we want to re-try the rollback a few times?